rfcs/01780_policy_delegation.md

```diff
@@ -0,0 +1,308 @@
+the one that is being verified to which the hash refers.
+
+Admittedly all of this is unlikely to be performed by regular (or even advanced)
+users. Also, it would be unlikely to be useful in practice, because code often
+evolves, and therefore previous data labelled with a specific Wasm module hash
+label would not be able to be declassified by new versions of the code, whatever
+that means in practice (e.g. a version with bug fixes or improved
+functionality).
+
+### Wasm Module Signature
+
+https://github.com/project-oak/oak/blob/573e61e74668db3eb53747f38757b00aa899f1f6/oak_abi/proto/label.proto#L71-L77
+
+In order to facilitate more useful data flows with declassification other than
+just individual Wasm modules identified by a precise hash, Oak also supports
+Wasm module signature labels.
+
+A Wasm module signature principal refers to a public key that is used to sign
+any number of individual Wasm modules (or more precisely, to sign their hashes).
+
+For instance, let's assume there are three Wasm modules, with hashes
+respectively w_0, w_1, w_2. Also there are two (public, secret) key pairs (p_0,
+s_0), (p_1, s_1).
+
+Let's assume that the modules are supposed to implement some declassification
+logic, e.g. counting the number of people in an image given raw pixels. s_0 and
+s_1 are two parties that verify such models (from correctness, privacy and
+security point of view) and sign them if they (respectively) believe that they
+are trustworthy.
+
+The owner of s_0 signs w_0 and w_1 with s_0, while the owner of s_1 signs w_1
+and w_2 with s_1:
+
+```text
+    ┌ w_0
+s_0 │
+    └ w_1 ┐
+          │ s_1
+      w_2 ┘
+```
+
+## Delegation Manifest
+
+The main contribution of this RFC is to specify a mechanism by which parties may
+publish a list of hashes in a standard format and serve it statically over
+HTTPS.
+
+In the example above, the owner of s_0 would need to publish the signed module
+hashes somewhere, and a client would have to know the value of s_0 itself (the
+actual key used for signing the module), which introduces extra complexity and
+parties around the deployment and publishing of new modules.
+
+Here we introduce a delegation mechanism that is based on the publication of a
+manifest file on an HTTPS endpoint.
+
+A Delegation Manifest is identified (and located) by an HTTPS URL pointing to a
+JSON file with the schema described below.
+
+For instance, Google may publish a delegation manifest under
+`https://google.com/oak/manifests/count_people_from_pixels.json` with the
+following contents:
+
+```json
+{
+  "trusted": [
+    {
+      "type": "manifest",
+      "value": "https://eff.com/oak/manifests/count_people_from_pixels.json"
+    },
+    {
+      "type": "wasm_module",
+      "value": "65c296d60c077db03c7ec90c60107a52221c14bb"
+    },
+    {
+      "type": "wasm_module",
+      "value": "cf26b96746311908f8fe40340a8b6c1f06f5a1b5"
+    }
+  ]
+}
```

Comment:
Is there a signature associated with the json file? If so is it included in the json or separately?

Reply:
In this version, there are no keys or signatures. I am intentionally sticking to TLS only, since it's the only PKI that is already deployed and we know it works. Additional or orthogonal PKIs may be considered as future work, but that would require setting up an entire ecosystem similar to TLS just to share those keys. (see cargo-crev and the problem of bootstrapping trust there). In practice probably any other kind of key would still be fetched over TLS at least initially, so this seems a reasonable compromise for now.

Reply:
If a specific company (or user) becomes popular and has many people trusting their manifests over many different Oak instances, that comapny (or user) becomes a very attractive target for attackers.

In the case of normal web sites, an invalid certificate usually affectes only one site. In this case the blast radius can be much wider and much more difficult to find the affected users to notify them as they may be spread over many different platforms.

Reply:
That's correct, and similar to highly popular code dependencies (e.g. absl, various web frameworks, etc.). In principle github or wherever the code is hosted would be able to single-handedly serve backdoored code, or an attacker may compromise one of the admin account of that repository. The same rules would apply here for the delegation manifests, in terms of allowing edit access, and generally having best practices around editing it. WDYT?